service/src/main/antlr4/org/ehrbase/aql/parser/Aql.g4

```diff
@@ -111,24 +111,24 @@ contains
 
 identifiedExpr
         : identifiedEquality ((OR|XOR|AND) identifiedEquality)*
-        | OPEN_PAR identifiedEquality ((OR|XOR|AND) identifiedEquality)* CLOSE_PAR
+        | OPEN_PAR identifiedEquality ((OR|XOR|AND)identifiedEquality)* CLOSE_PAR
         ;
 
 //identifiedExprAnd
 //      : identifiedEquality (AND identifiedEquality)*;
 //      : identifiedEquality (AND identifiedExpr)*;
 
 identifiedEquality
-        : NOT? identifiedOperand COMPARABLEOPERATOR identifiedOperand
-        | NOT? identifiedOperand MATCHES OPEN_CURLY matchesOperand CLOSE_CURLY
-        | NOT? identifiedOperand MATCHES REGEXPATTERN
-        | NOT? identifiedOperand LIKE STRING
-        | NOT? identifiedOperand ILIKE STRING
-        | NOT? identifiedOperand SIMILARTO STRING
+        : OPEN_PAR? NOT? identifiedOperand COMPARABLEOPERATOR identifiedOperand CLOSE_PAR?
+        | OPEN_PAR? NOT? identifiedOperand MATCHES OPEN_CURLY matchesOperand CLOSE_CURLY CLOSE_PAR?
+        | OPEN_PAR? NOT? identifiedOperand MATCHES REGEXPATTERN CLOSE_PAR?
+        | OPEN_PAR? NOT? identifiedOperand LIKE STRING CLOSE_PAR?
+        | OPEN_PAR? NOT? identifiedOperand ILIKE STRING CLOSE_PAR?
+        | OPEN_PAR? NOT? identifiedOperand SIMILARTO STRING CLOSE_PAR?
```

Comment:
Just an idea of mine, as I am a fan of more expressive grammars: Instead of modifying the rules, add additional rules including the parenthesis. This would ensure by grammar side that the parenthesis will always be balanced, instead of building your own PDA.  

```suggestion
        NOT? identifiedOperand COMPARABLEOPERATOR identifiedOperand
        | OPEN_PAR NOT? identifiedOperand COMPARABLEOPERATOR identifiedOperand CLOSE_PAR
        | NOT? identifiedOperand MATCHES OPEN_CURLY matchesOperand CLOSE_CURLY
        | OPEN_PAR NOT? identifiedOperand MATCHES OPEN_CURLY matchesOperand CLOSE_CURLY CLOSE_PAR
        | NOT? identifiedOperand MATCHES REGEXPATTERN
        | OPEN_PAR NOT? identifiedOperand MATCHES REGEXPATTERN CLOSE_PAR
        | NOT? identifiedOperand LIKE STRING
        | OPEN_PAR NOT? identifiedOperand LIKE STRING CLOSE_PAR
        | NOT? identifiedOperand ILIKE STRING
        | OPEN_PAR NOT? identifiedOperand ILIKE STRING CLOSE_PAR
        | NOT? identifiedOperand SIMILARTO STRING
        | OPEN_PAR NOT? identifiedOperand SIMILARTO STRING CLOSE_PAR
```

Reply:
Thanks Niklas, unfortunately, we have to support multiple levels of enclosing parenthesis (as raised by Birger). I modified the grammar to support this.

Reply:
Yeahn, then it makes sense to implement it that way! 👍